collide: fix collision issues with bridges, drawbridges, and trapdoors

walkawayy opened the pull request:
#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description
Fixed collision issues with drawbridges, trapdoors, and bridges when stacked and over slopes.

Attached @lahm86's test level that has a huge number of scenarios to test with.
[Bridges Test Level.zip](https://github.com/user-attachments/files/16420855/Bridges.Test.Level.zip)
EDIT: More extensive test level: [LEVEL1.zip](https://github.com/user-attachments/files/16426768/LEVEL1.zip)

walkawayy committed:
types: rename Lara state LS_NULL to LS_CLIMB_UP

walkawayy commented:
I left some TODO comments in the PR for some questions I had. I also will clean up some of the files where functions changed from static, but I didn't want to blow up the diff for the PR. I also am expecting some discussion on the code changes, so I figure the usage might change depending on the code review anyway.

rr- reviewed (commented):
Any chance we can make these improvements a part of the `fix_bridge_collision` setting?

rr- commented on src/game/collide.c:
```diff
@@ -40,10 +43,27 @@ void Collide_GetCollisionInfo(
+
+    // TODO Do we need to check all walkable heights or is knowing Lara is above
+    // one bridge enough? Can a slope ever be below one bridge but above
+    // another?
```
I don't think this is possible unless a level builder embeds an object in the room geometry, which serves no purpose. Or they embed the bridge in a way so that it protrudes from a slope and is to use mixed room/object collision, but it's unreasonable to expect that to work IMO.
So all in all I think it's good to remove this TODO.

rr- commented on src/game/collide.c:
```diff
@@ -40,10 +43,27 @@ void Collide_GetCollisionInfo(
+    for (int32_t i = 0; i < walkable_heights->count; i++) {
+        int32_t item_height = (int32_t)walkable_heights->items[i];
+        // TODO Need check within WALL_L range? Can walkables ever be below a
+        // slope?
```
It can't unless of course we talk about sloped ceilings.

rr- commented on src/game/collide.c:
```diff
@@ -40,10 +43,27 @@ void Collide_GetCollisionInfo(
+        // slope?
+        if (ABS(item_height - ypos) < WALL_L
+            && ABS(item_height - ypos) <= STEPUP_HEIGHT) {
+            skip_slope_check = true;
```
We can `break;` here as the only thing we do in this loop is set this bool.

rr- commented on src/game/collide.c:
```diff
@@ -40,10 +43,27 @@ void Collide_GetCollisionInfo(
+        coll->tilt_z = tilt >> 8;
+        coll->tilt_x = (int8_t)tilt;
+    }
+    Vector_Clear(walkable_heights);
```
This should've been `Vector_Free`, since we'll later override the content of this variable with a separately allocated vector.

rr- commented on src/game/collide.c:
```diff
@@ -114,19 +135,32 @@ void Collide_GetCollisionInfo(
+    skip_slope_check = false;
+    walkable_heights = Room_GetWalkableHeights(floor, x, ypos, z);
+    for (int32_t i = 0; i < walkable_heights->count; i++) {
+        int32_t item_height = (int32_t)walkable_heights->items[i];
```
`const int32_t` (but only in the variable type, not in the cast, as making primitive casts const serves no purpose)

rr- commented on src/game/collide.c:
```diff
@@ -114,19 +135,32 @@ void Collide_GetCollisionInfo(
+    for (int32_t i = 0; i < walkable_heights->count; i++) {
+        int32_t item_height = (int32_t)walkable_heights->items[i];
+        if (ABS(item_height - ypos) <= STEPUP_HEIGHT) {
+            skip_slope_check = true;
```
`break;`

rr- commented on src/game/collide.c:
```diff
@@ -114,19 +135,32 @@ void Collide_GetCollisionInfo(
+            coll->front_floor = 512;
+        }
     }
+    Vector_Clear(walkable_heights);
```
`Vector_Free`

rr- commented on src/game/collide.c:
```diff
@@ -143,19 +177,32 @@ void Collide_GetCollisionInfo(
+    skip_slope_check = false;
+    walkable_heights = Room_GetWalkableHeights(floor, x, ypos, z);
+    for (int32_t i = 0; i < walkable_heights->count; i++) {
+        int32_t item_height = (int32_t)walkable_heights->items[i];
```
`const int32_t`

rr- commented on src/game/collide.c:
```diff
@@ -143,19 +177,32 @@ void Collide_GetCollisionInfo(
+            coll->left_floor = 512;
+        }
     }
+    Vector_Clear(walkable_heights);
```
`Vector_Free`

rr- commented on src/game/collide.c:
```diff
@@ -172,18 +219,30 @@ void Collide_GetCollisionInfo(
+    skip_slope_check = false;
+    walkable_heights = Room_GetWalkableHeights(floor, x, ypos, z);
+    for (int32_t i = 0; i < walkable_heights->count; i++) {
+        int32_t item_height = (int32_t)walkable_heights->items[i];
```
`const int32_t`

rr- commented on src/game/collide.c:
```diff
@@ -172,18 +219,30 @@ void Collide_GetCollisionInfo(
+    for (int32_t i = 0; i < walkable_heights->count; i++) {
+        int32_t item_height = (int32_t)walkable_heights->items[i];
+        if (ABS(item_height - ypos) <= STEPUP_HEIGHT) {
+            skip_slope_check = true;
```
`break;`

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -163,13 +177,16 @@ void Bridge_DrawBridgeCeiling(
+        return;
     }
+
+    *height = (int16_t)item->pos.y + STEP_L;
```
I don't think this cast is necessary as the variable type is already `int16_t`

walkawayy replied:
`item->pos.x/y/z` are `int32_t` I believe. Is the cast unneeded? There are some mismatched types around the codebase similar to this it seems. I just changed this since I was already in the relevant code.

rr- replied:
Most often yes, type promotion is implicit.

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -197,67 +215,73 @@ void Bridge_FlatFloor(
+        return;
     }
+
+    *height = (int16_t)item->pos.y + STEP_L;
```
Unnecessary cast

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -197,67 +215,73 @@ void Bridge_FlatFloor(
 
-    int32_t level = item->pos.y + (Bridge_GetOffset(item, x, y, z) >> 2);
-    if (y > level) {
+    int32_t offset_height = item->pos.y + (Bridge_GetOffset(item, x, y, z));
```
`const int32_t`

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -197,67 +215,73 @@ void Bridge_FlatFloor(
-    int32_t level = item->pos.y + (Bridge_GetOffset(item, x, y, z) >> 2);
-    if (y > level) {
-        *height = level + STEP_L;
+    int32_t offset_height = item->pos.y + (Bridge_GetOffset(item, x, y, z));
```
`const int32_t`

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -197,67 +215,73 @@ void Bridge_FlatFloor(
+        return;
     }
+
+    *height = (int16_t)offset_height + STEP_L;
```
Unnecessary cast

walkawayy replied:
`height` is `int16_t` while `offset_height` is `int32_t`. Should this have a cast?

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -197,67 +215,73 @@ void Bridge_FlatFloor(
 
-    int32_t level = item->pos.y + (Bridge_GetOffset(item, x, y, z) >> 1);
-    if (y > level) {
+    int32_t offset_height = item->pos.y + (Bridge_GetOffset(item, x, y, z));
```
`const int32_t`

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -197,67 +215,73 @@ void Bridge_FlatFloor(
-    int32_t level = item->pos.y + (Bridge_GetOffset(item, x, y, z) >> 1);
-    if (y > level) {
-        *height = level + STEP_L;
+    int32_t offset_height = item->pos.y + (Bridge_GetOffset(item, x, y, z));
```
`const int32_t`

rr- commented on src/game/objects/general/bridge.c:
```diff
@@ -197,67 +215,73 @@ void Bridge_FlatFloor(
+        return;
     }
+
+    *height = (int16_t)offset_height + STEP_L;
```
Unnecessary cast

rr- commented on src/game/objects/general/bridge.h:
```diff
@@ -4,6 +4,8 @@
 #include <stdint.h>
 
+int32_t Bridge_GetOffset(ITEM_INFO *item, int32_t x, int32_t y, int32_t z);
+bool Bridge_OnDrawBridge(ITEM_INFO *item, int32_t x, int32_t z);
```
I don't think we ever use `Bridge_OnDrawBridge` publicly, so this can be made `static` back.

rr- commented on src/game/objects/general/trapdoor.h:
```diff
@@ -4,6 +4,7 @@
 
 #include <stdint.h>
 
+bool TrapDoor_StandingOn(ITEM_INFO *item, int32_t x, int32_t z);
```
I don't think we ever use `TrapDoor_StandingOn` publicly, so it can be made `static` back.

rr- commented on src/game/objects/general/trapdoor.c:
```diff
@@ -65,6 +67,8 @@ void TrapDoor_Ceiling(
         return;
     }
+
+    // TODO Invert condition?
```
Yes, to be consistent with the bridge code change.

rr- commented on src/game/objects/general/trapdoor.c:
```diff
@@ -53,9 +53,11 @@ void TrapDoor_Floor(
         return;
     }
+
+    // TODO Invert condition?
```
Yes, to be consistent with the bridge code change.

rr- commented on src/game/objects/general/trapdoor.c:
```diff
@@ -53,9 +53,11 @@ void TrapDoor_Floor(
     if (y <= item->pos.y && item->current_anim_state == DOOR_CLOSED
         && item->pos.y < *height) {
-        *height = item->pos.y;
+        *height = (int16_t)item->pos.y;
```
Unnecessary cast

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+VECTOR *Room_GetWalkableHeights(
+    FLOOR_INFO *floor, int32_t x, int32_t y, int32_t z)
+{
+    ROOM_INFO *r;
```
Double declaration of `r`

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+{
+    ROOM_INFO *r;
+    while (floor->pit_room != NO_ROOM) {
+        ROOM_INFO *r = &g_RoomInfo[floor->pit_room];
```
`const ROOM_INFO *const r`

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+    while (floor->pit_room != NO_ROOM) {
+        ROOM_INFO *r = &g_RoomInfo[floor->pit_room];
+        int32_t x_floor = (z - r->z) >> WALL_SHIFT;
+        int32_t z_floor = (x - r->x) >> WALL_SHIFT;
```
`const int32_t` for both values

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+                    if (TRIG_BITS(trigger) == TO_CAMERA) {
+                        trigger = *floor_data++;
+                    }
+                } else {
```
This can be rearranged to

````c
if (TRIG_BITS(trigger == TO_OBJECT) {
    int16_t item_num = ...
    ...
} else if (TRIG_BITS(trigger) == TO_CAMERA) {
    trigger = *floor_data++;
}
```

walkawayy replied:
Ok. I reused code from the GetFloor and GetHeight functions here which do it like this.

I had a wish to have a common floor traversal function since they are all slightly different and repeated, but each time it happens there are different things being done while traversing. Stats sets killables, secrets, and pickups, injection does its own thing, etc.

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+                    ITEM_INFO *item = &g_Items[item_num];
+                    if (Object_IsObjectType(
+                            item->object_number, g_TrapdoorObjects)) {
+                        int32_t item_height = item->pos.y;
```
`const int32_t`

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+                    }
+                } else {
+                    int16_t item_num = trigger & VALUE_BITS;
+                    ITEM_INFO *item = &g_Items[item_num];
```
`const ITEM_INFO *const`

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+                        trigger = *floor_data++;
+                    }
+                } else {
+                    int16_t item_num = trigger & VALUE_BITS;
```
`const int16_t`

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+                        Vector_Add(heights, (void *)item->pos.y);
+                    } else if (Object_IsObjectType(
+                                   item->object_number, g_BridgeObjects)) {
+                        int32_t item_height =
```
`const int32_t item_height`

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
 }
+
+VECTOR *Room_GetWalkableHeights(
+    FLOOR_INFO *floor, int32_t x, int32_t y, int32_t z)
```
`const FLOOR_INFO *const floor, const int32_t x, const int32_t y, const int32_t z`

rr- commented on src/game/room.h:
```diff
@@ -30,3 +32,5 @@ void Room_AlterFloorHeight(ITEM_INFO *item, int32_t height);
 void Room_TestTriggers(int16_t *data, bool heavy);
 void Room_FlipMap(void);
+VECTOR *Room_GetWalkableHeights(
+    FLOOR_INFO *floor, int32_t x, int32_t y, int32_t z);
```
```c
VECTOR *Room_GetWalkableHeights(
    const FLOOR_INFO *floor, int32_t x, int32_t y, int32_t z);
```

rr- commented on src/game/room.c:
```diff
@@ -977,3 +981,66 @@ void Room_TestTriggers(int16_t *data, bool heavy)
+    } while (!(type & END_BIT));
+
+    return heights;
+}
```
Rather than consulting bridge code directly from room which is a big no-no, we should introduce a new function pointer in OBJECT_INFO, for example, `get_walkable_height`, that we'd set to `Bridge_GetOffset` in relevant `…_Setup` functions for the `g_BridgeObjects` members, and call it here via `object->get_walkable_height`. This way we'd probably no longed need `g_BridgeObjects`.

walkawayy replied:
I will double check, but the ceiling and floor fixes for bridges I made last. I wonder if those are fixed, if we can simplify this part of the code somehow.

rr- reviewed (commented):
Any chance we can make these improvements a part of the `fix_bridge_collision` setting?

---

When reviewing `Room_GetWalkableHeights` I noticed that we seem to have already something in the code that sounds like it serves the same purpose – namely, `OBJECT_INFO`'s `void (*floor)(ITEM_INFO *item, int32_t x, int32_t y, int32_t z, int16_t *height);` used by `Room_GetHeight`. The slope checks also call `Room_GetHeight`. So, is there any chance we could bake the introduced checks into `Room_GetHeight`?

rr- commented on src/game/collide.c:
```diff
@@ -40,10 +43,27 @@ void Collide_GetCollisionInfo(
+        int16_t tilt = Room_GetTiltType(floor, x, g_LaraItem->pos.y, z);
+        coll->tilt_z = tilt >> 8;
+        coll->tilt_x = (int8_t)tilt;
+    }
```
I think we should reset the tilt values back to zero in case we deem Lara is not a slope. Without that, I think it might be the case that when Lara slides into a bridge, she'll continue sliding.

lahm86 replied:
Test updates attached.
[LEVEL1.zip](https://github.com/user-attachments/files/16426768/LEVEL1.zip)

walkawayy replied:
This actually fixed a small bug in Lahm's newest test level where Lara would slide for a few frames when walking over tilted bridges when slightly above a slope. Nice catch.

rr- reviewed (commented):
One last question: what will happen if we only use the first `skip_floor_check` and let the others (front/left/right) stay as they were? I think these other checks are mostly about preventing Lara from traversing up or down the slopes depending on her animations.

Also upon reading the code it'd be good to have another test in @lahm86 room:
- A slope and a slightly tilted floor that directly continues onto a bridge/trap door
- Same as above but with a slight positive or negative height difference
- Various geometries for checking if Lara can or cannot fit underneath a bridge or a trapdoor

walkawayy commented:
I will address the smaller comments as I incorporate changes and test the code. I am going to introduce a separate commit for the trapdoor and drawbridge cleanups / condition inversions. I was just in that code a lot as I was investigating this issue, so that's why I even bothered with them.

I agree linking these fixes to `fix_bridge_collision` will be nice. I just need to think of a clean way to do this since there's a bunch of difference code affected.

As far as incorporating `Room_GetWalkableHeights`into `Room_GetHeight`, that's possibly a good idea. I just was hesitant to make any big changes in fear of changing the game collision behavior. Especially as we will need to account for future TRs which have slight changes as well. I guess if we can make it configurable with the option, maybe that can be more cleanly done. TBD on if it will work.

walkawayy commented:
`what will happen if we only use the first skip_floor_check and let the others (front/left/right) stay as they were? I think these other checks are mostly about preventing Lara from traversing up or down the slopes depending on her animation`

This change completely breaks Lara's movement. Lara will be pushed away or invisible walled from any bridge over a tilt as her collision views the movement as blocked.

walkawayy commented:
Ok so now that the bridges' floor functions are fixed, I believe I can simplify the skip slope check by caching the room height like this:
```
    int32_t height = Room_GetHeight(floor, x, ytop, z);
    const int32_t room_height = height;
```

And then checking:
```
        if (room_height == ypos) {
            skip_slope_check = true;
            break;
        }
```

However, I still need to add an `||` to that condition to also check if the floor is equal to a walkable height. I need to know if the positions are equal because Lara is on a walkable item or just equal because Lara is standing on a normal floor. So, I still think I need to traverse the walkable items on the floor tile and get their heights. The only other solution I see is if we set a global variable in `Room_GetHeight` whenever a `TO_OBJECT` is reached and its floor function runs:
```
if (object->floor) {
    object->floor(item, x, y, z, &height);
}
```

I'm not sure if we want to do that, but I will propose it. Maybe there's a cleaner way to use that information somehow. Also, I did just notice `g_TriggerIndex` is a variable in there that could at least simplify `Room_GetWalkableHeights` since we could just check the trigger data instead of having to traverse the floor again. It's a small optimization. I believe that's what it contains anyway.

rr- commented:
I'm seeing most of the comments are resolved, but there are no corresponding changes – can I ask to push what we have so far for easier bookkeeping?

walkawayy committed:
trapdoor: invert floor and ceiling if conditions

walkawayy committed:
bridge: make Bridge_OnDrawBridge similar to TrapDoor_StandingOn

walkawayy committed:
collide: fix collision issues with bridges

Fixed collision issues with drawbridges, trapdoors, and bridges when
stacked over each other, over slopes, and near the ground.

Resolves #606.

walkawayy commented:
The diff is best viewed commit by commit. The first few are just minor cleanups. I tested Lahm's level with the bridge fix config option enabled and disabled. With it disabled, OG behavior seemed preserved (and broken). I think this solves just about all remaining issues we saw. There are only a few small outliers where Lara can land specific jumps into a tight space and sometimes bug on top of a bridge similar to the corner bug if a bridge is placed extremely low over a high slope. But I don't think that is a realistic level design and Lara can jump normally on top of that bridge anyway. I think it's also just the original engine collision. I didn't want to mess with the game fundamentals too much with this. LMK what you think.

lahm86 commented on src/game/room.c:
```diff
@@ -13,7 +15,9 @@
 
+#include <libtrx/log.h>
 #include <libtrx/utils.h>
+#include <libtrx/vector.h>
```
I think this is redundant now?

lahm86 replied:
I think this is redundant now? And the log include too.

lahm86 commented on src/game/room.h:
```diff
@@ -3,6 +3,8 @@
 #include "global/const.h"
 #include "global/types.h"
 
+#include <libtrx/vector.h>
```
And here.

lahm86 commented on tools/config/TR1X_ConfigTool/Resources/Lang/en.json:
```diff
@@ -155,7 +155,7 @@
     "fix_bridge_collision": {
       "Title": "Fix bridge collision",
-      "Description": "Fixes Lara not being able to grab parts of some bridges and invisible walls at the edge."
+      "Description": "Fixes Lara not being able to grab parts of some bridges and invisible walls at the edge. Also fixes issues collision issues with drawbridges, trapdoors, and bridges when stacked over each other, over slopes, and near the ground."
```
`...fixes issues collision issues...`

lahm86 reviewed (commented):
Just some minor comments, but otherwise everything looks 👍

rr- reviewed (approved):
The end result has become very elegant IMO.

rr- reviewed (changes requested):
Hold on – please remove the subprojects/libtrx change.

walkawayy commented:
@rr- I think I fixed the submodule issue. If it looks good, feel free to merge.
